packages/components/pie-button/src/decorators.ts

```diff
@@ -0,0 +1,34 @@
+/**
+ * A decorator for specifying a list of valid values for a property.
+ * If this property's setter is called with an invalid value, an error is logged and the default value will be assigned instead.
+ * @param validValues - The array of valid values
+ * @param defaultValue - The value to fall back on
+ * @returns
+ */
+export const validValues = (validValues: any[], defaultValue: any) => {
```

Comment:
Curious to look at some reference material supporting this method for Lit prop validation, if you may have some at hand 🤔

Reply:
Sure, [Lit suggests](https://lit.dev/docs/components/properties/#accessors-custom) adding custom getters/setters for stuff like validation and fallback value handling. An example of this in the wild can be seen in [Spectrum](https://github.com/adobe/spectrum-web-components/blob/007a762c0257af4cd74e3781d877a8a8343acfd0/packages/button/src/Button.ts#L57). Our opinion is this is messy and leads to quite bloated classes. Personally I think the decorator is far more elegant easier to understand but love to hear other people's thoughts!

Reply:
Hey, the Lit docs don't really mention property validation, and we had a look around at some of the other design systems and how they're using web components, but all of those seemed to have a lot of duplication.

But we thought that validation was required for a few reasons:

1. If someone enters an incorrect value (e.g., `size="xlarge"` which is not supported), we need to know what to do in that situation (fall back to a default value), and also to let the engineer know that they are using a bad value (via an error in the console).
2. To fulfil the first requirement, we need to be able to say which values are valid or not.

In the button component there are at least 3 properties (size, style and icon position) that require validation, and there will be plenty of examples in other components too. Most other people who are doing this seem to be writing a `get` and `set` for each property which is not very DRY and is also bad for readability in the component. Having this decorator not only keeps the logic reusable (we'll extract it out of the button at some point) but also lightweight in the component file.

I'll be doing a demo of this to the team soon, but if you have any more questions in the mean time then feel free to message me or I can show you one-on-one!

Reply:
The decorator solution looks super nice. Good job!